### ISSUE TITLE
After fine-tuning, the model repeat the answer and not stop 

### ISSUE THREAD

**Issue by abdoelsayed2016** [tloen/alpaca-lora#467]
After fine-tuning, I have encountered an unexpected behavior where the model repeatedly generates the same answer without stopping.

i tried different models of llama 

decapoda-research/llama-7b-hf 
yahma/llama-7b-hf
here are some examples 
 ```
 Tell me about alpacas.
 
 
 1. Alpacas are from South America.
2. Alpacas are from South America.
3. Alpacas are from South America.
4. Alpacas are from South America.
5. Alpacas are from South America.
6. Alpacas are from South America.
7. Alpacas are from South America.
8. Alpacas are from South America.
9. Alpacas are from South America.
10. Alpacas are from South America.
11. Alpac
```

```
who played charlie bucket in the original charlie and the chocolate factory

\begin{itemize}
\item Freddie Highmore
\item Freddie Highmore
\item Freddie Highmore
\item Freddie Highmore
\item Freddie Highmore
\item Freddie Highmore
\item Freddie Highmore
\item Freddie Highmore
\item Freddie Highmore
\item Freddie Highmore
\item Freddie Highmore
\item Freddie Highmore
\item Freddie Highmore
\item Freddie Highmore
\item Freddie Highmore
\item Freddie Highmore
\item Freddie Highmore
\
```

Also, adapter_model.bin size is 1kb

**Comment by aryanq123** [comment]
can someone assign me this task i wants to work upon it

and explain briefly

**Comment by Qubitium** [comment]
@abdoelsayed2016 You using an old model that is no longer recommended as it was convertes to hf using wrong tokenizer params. Check the repo readme for valid 7b hf models.

**Comment by abdoelsayed2016** [comment]
@diegomontoya I have already read the readme many times. I see that while finetuning the base model, is` decapoda-research/llama-7b-hf`
**I have two questions. First, why does the model repeat the answer? The model doesn't know when it should stop** 
![image](https://github.com/tloen/alpaca-lora/assets/27821589/13310b96-6d29-4c25-87ce-63b187b8cf85)


**The second question is whether the adapter weight is 1KB only. Is that normal?**

![image](https://github.com/tloen/alpaca-lora/assets/27821589/59c774e4-fd3c-42fd-8386-0f3ccdf1f6b7)

**Comment by abdoelsayed2016** [comment]
Also i tired `huggyllama/llama-7b`

**Comment by abdoelsayed2016** [comment]
I see a lot of issues talking about the size of the adapter I tried all solutions, but still, the adapter is the same size 1KB.

anyone has faced the same issue cuz training is costly

**Comment by asenmitrev** [comment]
I was able to solve this issue with this [comment](https://github.com/tloen/alpaca-lora/issues/334#issuecomment-1509599096). Uninstall peft and install from github commit.

I was desperate with the repetitions for both models (yahma and decapoda-research) but I think the root cause is the empty adapter model.bin.

If you want to do a quick test without training, take the last checkpoint `pytorch_model.bin` and replace the `adapter_model.bin` with that, renaming it. This allowed me to test the model and the repetition was fixed.

**Comment by abdoelsayed2016** [comment]
@asenmitrev

Yesterday I did this solution but still adapter model.bin size 1KB 
```

pip uninstall peft -y
pip install git+https://github.com/huggingface/peft.git@e536616888d51b453ed354a6f1e243fecb02ea08
```

i tried again with uninstall peft. Here is the output 

![image](https://github.com/tloen/alpaca-lora/assets/27821589/e9c22fe4-2f67-4a27-8d9c-d46d3d8b3ce8)

**Comment by abdoelsayed2016** [comment]
I also tried this solution 
https://github.com/huggingface/peft/issues/286
https://github.com/tloen/alpaca-lora/issues/37#issuecomment-1473140882

The adapter size now is 7 GB 

And model back to repeat the answers.

![image](https://github.com/tloen/alpaca-lora/assets/27821589/07a5b6e2-e6e7-4776-8d57-2a5bf2edfb0d)


I also changed the checkpoint model name  from pytorch_model.bin to the adapter_model.bin


same output

**Comment by asenmitrev** [comment]
I tried decapoda research and the answer is repeated. yahma is fine

**Comment by abdoelsayed2016** [comment]
@asenmitrev Yahma with which version of peft? 
Because i tired it same problem
I got confused 
They should add description in readme or fixed requirements

**Comment by asenmitrev** [comment]
The one quoted in my first answer.

**Comment by SimpleConjugate** [comment]
You should try adding `repetition_penalty` keyword argument to generation config in the evaluate function.  `repetition_penalty >1` should do it. I finetuned a model and used `repetition_penalty=2` to resolve the problem for myself.

**Comment by SuperBruceJia** [comment]
## Issue
We also encountered this recurring response issue while training the **GPTQ-quantized models with a LoRA adapter**. We met the same problem while fine-tuning the LoRA adapter with the GPTQ-quantized [Command R Plus](https://huggingface.co/CohereForAI/c4ai-command-r-plus) ([our model](https://huggingface.co/shuyuej/Command-R-Plus-GPTQ)) and GPTQ-quantized [LLaMA 3.3 70B](https://huggingface.co/meta-llama/Llama-3.3-70B-Instruct) ([our model](https://huggingface.co/shuyuej/Llama-3.3-70B-Instruct-GPTQ)). **After the model quantization using the GPTQ algorithm, the model performs pretty well without noticeable recurring response issues! However, after we fine-tuned the LoRA adapter, things are getting tricky.** We found that the problem came from **increasing training loss** and **increasing gradient norm**.  

### Training Loss Before Using Our Solution
Please check our training loss and gradient norm figures.
![image](https://github.com/user-attachments/assets/0f152e12-cfc2-4fdd-8e61-8a5c65d14368)

### Repeated Response Examples
Here are the responses from the quantized model + trained LoRA:
![image](https://github.com/user-attachments/assets/8943a7a5-616e-45f7-8a6c-c6b8886e8a84)
![image](https://github.com/user-attachments/assets/5d6c75f0-e4c0-47ed-8213-cff9c04917e9)
![image](https://github.com/user-attachments/assets/d960b56a-c5b1-42e6-8e58-e1532bf083be)
![image](https://github.com/user-attachments/assets/d9329b80-2a19-428f-9feb-5efb67cd1f04)

## The Reason to This Issue and Our Solution
We believe that the **increasing training loss** issue was from the **Gradient Exploding**. So, we made changes to our [hyperparameter settings](https://github.com/vkola-lab/PodGPT/blob/main/config_quantization.yml#L40-L68). Also, please check our [model loader](https://github.com/vkola-lab/PodGPT/blob/main/lib/model_loader_quantization.py) if you are interested. Based on these settings, we solved this issue and **achieved stable training**:

### Training Loss After Using Our Solution
![quantized_model_training_progress](https://github.com/user-attachments/assets/9de46d19-2243-4a73-b908-3442e2cccbb8)

## Model Inference
**Based on our model's responses, we believe we have fundamentally solved the issue of repeated responses!**
<img width="2053" alt="image" src="https://github.com/user-attachments/assets/f77f28d0-2286-487d-9be9-6179c8844c76" />
<img width="2053" alt="image" src="https://github.com/user-attachments/assets/bde9750a-f1e0-4359-958b-84b290b9e113" />
<img width="2052" alt="image" src="https://github.com/user-attachments/assets/85f1bcec-b937-417b-bf43-953888a477fb" />


Best regards,

Shuyue
Dec. 15th, 2024
